app/controllers/registers_controller.rb

```diff
@@ -57,36 +57,64 @@ def show
 
   def history
     @register = Spina::Register.find_by_slug!(params[:id])
-    @register_data = @registers_client.get_register(@register.name.parameterize, @register.register_phase)
-    fields = @register_data.get_field_definitions.map { |field| field.item.value['field'] }
-    all_records = @register_data.get_records_with_history
-    entries_reversed = @register_data.get_entries.reverse_each
 
-    entries_mapped_with_items = entries_reversed.map do |entry|
-      records_for_key = all_records.get_records_for_key(entry.key)
-      current_record = records_for_key.detect { |record| record.entry.entry_number == entry.entry_number }
-      previous_record_index = records_for_key.find_index(current_record) - 1
+    unless @register.populated
+      populate_register(@register)
+    end
+
+    entries = recover_entries_history(@register.id, @register.fields, params)
+    register_data = @registers_client.get_register(@register.name.parameterize, @register.register_phase)
+    fields = register_data.get_field_definitions.map { |field| field.item.value['field'] }
 
-      if previous_record_index.negative?
+    @entries_with_items = entries.map do |entry_history|
+      current_record = entry_history[:current_entry]
+
+      if entry_history[:previous_entry].nil?
         changed_fields = fields
       else
-        previous_record = records_for_key[previous_record_index]
-        changed_fields = fields.reject { |f| current_record.item.value[f] == previous_record.item.value[f] }
+        previous_record = entry_history[:previous_entry]
+        changed_fields = fields.reject { |f| entry_history[:current_entry].data[f] == previous_record.data[f] }
       end
 
-      { current_record: current_record, previous_record: previous_record, updated_fields: changed_fields, key: entry.key }
+      { current_record: current_record, previous_record: previous_record, updated_fields: changed_fields, key: current_record.key }
     end
 
-    if params[:q].present?
-      filtered = filter(entries_mapped_with_items, params[:q])
-      @entries_with_items = paginate_history(filtered)
-    else
-      @entries_with_items = paginate_history(entries_mapped_with_items)
-    end
+    @entries_with_items = Kaminari::paginate_array(@entries_with_items, total_count: @page_count).page(@current_page).per(100)
   end
 
   private
 
+  def recover_entries_history(register_id, fields, params)
+    literal = params[:q]
+    page = if params[:page].nil? then 1 else params[:page].to_i end
+
+    if literal.present?
+      partial = ''
+      fields.split(',').each { |field| partial += " data->> '#{field}' ilike '%#{literal}%' or" }
+      partial = partial[1, partial.length - 3] # Remove beginning space and end 'or'
+
+      query = Entry.where("spina_register_id = '#{register_id}' and entry_type = 'user' and (#{partial})").order(:entry_number).reverse_order.limit(100).offset(100 * (page - 1))
```

Comment:
rather than having these `and` conditions. I think you can use the ORM to build this query e.g. 
`where("spina_register_id: register_id, entry_type: 'user')` etc

Reply:
Sure.